packages/instruction-plans/src/instruction-plan.ts

```diff
@@ -1 +1,526 @@
-export type InstructionPlan = unknown;
+import { Instruction } from '@solana/instructions';
+import {
+    appendTransactionMessageInstruction,
+    BaseTransactionMessage,
+    TransactionMessageWithFeePayer,
+} from '@solana/transaction-messages';
+import { getTransactionMessageSize, TRANSACTION_SIZE_LIMIT } from '@solana/transactions';
+
+/**
+ * A set of instructions with constraints on how they can be executed.
+ *
+ * This is structured as a recursive tree of plans in order to allow for
+ * parallel execution, sequential execution and combinations of both.
+ *
+ * Namely the following plans are supported:
+ * - {@link SingleInstructionPlan} - A plan that contains a single instruction.
+ *   This is a simple instruction wrapper and the simplest leaf in this tree.
+ * - {@link ParallelInstructionPlan} - A plan that contains other plans that
+ *   can be executed in parallel.
+ * - {@link SequentialInstructionPlan} - A plan that contains other plans that
+ *   must be executed sequentially. It also defines whether the plan is divisible
+ *   meaning that instructions inside it can be split into separate transactions.
+ * - {@link MessagePackerInstructionPlan} - A plan that can dynamically fill
+ *  instructions into transaction messages.
+ *
+ * Helpers are provided for each of these plans to make it easier to create them.
+ *
+ * @example
+ * ```ts
+ * const myInstructionPlan: InstructionPlan = parallelInstructionPlan([
+ *    sequentialInstructionPlan([instructionA, instructionB]),
+ *    instructionC,
+ *    instructionD,
+ * ]);
+ * ```
+ *
+ * @see {@link SingleInstructionPlan}
+ * @see {@link ParallelInstructionPlan}
+ * @see {@link SequentialInstructionPlan}
+ * @see {@link MessagePackerInstructionPlan}
+ */
+export type InstructionPlan =
+    | MessagePackerInstructionPlan
+    | ParallelInstructionPlan
+    | SequentialInstructionPlan
+    | SingleInstructionPlan;
+
+/**
+ * A plan wrapping other plans that must be executed sequentially.
+ *
+ * It also defines whether nested plans are divisible — meaning that
+ * the instructions inside them can be split into separate transactions.
+ * When `divisible` is `false`, the instructions inside the plan should
+ * all be executed atomicly — either in a single transaction or in a
+ * transaction bundle.
+ *
+ * You may use the {@link sequentialInstructionPlan} and {@link nonDivisibleSequentialInstructionPlan}
+ * helpers to create objects of this type.
+ *
+ * @example
+ * Simple sequential plan with two instructions.
+ * ```ts
+ * const plan = sequentialInstructionPlan([instructionA, instructionB]);
+ * plan satisfies SequentialInstructionPlan;
+ * ```
+ *
+ * @example
+ * Non-divisible sequential plan with two instructions.
+ * ```ts
+ * const plan = nonDivisibleSequentialInstructionPlan([instructionA, instructionB]);
+ * plan satisfies SequentialInstructionPlan & { divisible: false };
+ * ```
+ *
+ * @example
+ * Sequential plan with nested parallel plans.
```

Comment:
I'm pretty sure the way TypeDoc works is that you can name `@examples`.

This nit, everywhere it happens.

```suggestion
 * @example Simple sequential plan with two instructions.
 * ```ts
 * const plan = sequentialInstructionPlan([instructionA, instructionB]);
 * plan satisfies SequentialInstructionPlan;
 * ```
 *
 * @example Non-divisible sequential plan with two instructions.
 * ```ts
 * const plan = nonDivisibleSequentialInstructionPlan([instructionA, instructionB]);
 * plan satisfies SequentialInstructionPlan & { divisible: false };
 * ```
 *
 * @example Sequential plan with nested parallel plans.
```